packages/server/src/modules/step/service.ts

```diff
@@ -0,0 +1,108 @@
+import { Step, User } from '@chess-tent/models';
+import { DepopulatedStep, StepModel } from './model';
+import { AppDocument } from '@types';
+import { db, utils } from '@application';
+import { SubjectFilters } from '@chess-tent/types';
+import { FilterQuery } from 'mongoose';
+import _ from 'lodash';
+import { canAccessStepCheck, canEditStepCheck } from '@chess-tent/models/src';
+
+const depopulateStep = (
+  step: Partial<Step>,
+  depopulate = 'owner tags users',
+) => {
+  let transformed = new StepModel(step);
+  return transformed.depopulate(depopulate);
+};
+
+export const saveStep = (step: Step) =>
+  new Promise<void>(resolve => {
+    StepModel.updateOne({ _id: step.id }, depopulateStep(step), {
+      upsert: true,
+    }).exec(err => {
+      if (err) {
+        throw err;
+      }
+      resolve();
+    });
+  });
+
+export const getStep = (
+  stepId: Step['id'],
+  populate = 'owner tags users',
+): Promise<Step | null> =>
+  new Promise(resolve => {
+    StepModel.findById(stepId)
+      .populate(populate)
+      .exec((err, result) => {
+        if (err) {
+          throw err;
+        }
+        resolve(result ? result.toObject<Step>() : null);
+      });
+  });
+
+export const deleteStep = async (stepId: Step['id']) => {
+  await StepModel.deleteOne({ _id: stepId });
+};
+
+export const patchStep = (stepId: Step['id'], step: Partial<Step>) =>
+  new Promise<void>(resolve => {
+    StepModel.updateOne({ _id: stepId }, { $set: depopulateStep(step) }).exec(
+      err => {
+        if (err) {
+          throw err;
+        }
+        resolve();
+      },
+    );
+  });
+
+export const findSteps = (filters: Partial<SubjectFilters>): Promise<Step[]> =>
+  new Promise(resolve => {
+    const owner = utils.notNullOrUndefined({
+      owner: filters.owner,
+    });
+    const users = db.allQuery('users', filters.users);
+    const query: FilterQuery<AppDocument<DepopulatedStep>> =
+      utils.notNullOrUndefined({
+        published: filters.published,
+        ...db.orQueries(owner, users),
+      });
+
+    if (filters.difficulty) {
+      query['difficulty'] = { $eq: filters.difficulty };
+    }
+
+    if (!_.isEmpty(filters.tagIds)) {
+      // query['tags'] = { $in: filters.tagIds }; // todo: compiler is barking about thisxx
```

Comment:
???

Reply:
`NormalizedStep` doesn't have normalized `tags` field.

https://github.com/MrBr/chess-tent/pull/9/files/76e1e1a18417908df3640882458450f1f387b944#diff-1a70e3e6f97eb97d154e67f2c441402082d593b85138695cb97419493f9d2765R27

```
  tags: Step['tags'][]; // should be Tag['id'][];
```

Reply:
Barking is no more!